pkg/server/jwk_resolver.go

```diff
@@ -0,0 +1,127 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+
+	"github.com/abcxyz/pkg/cache"
+)
+
+type JWKResolver interface {
+	ResolveKeySet(ctx context.Context, oidcHeader string) (jwk.Set, error)
+}
+
+type OIDCResolver struct {
+	cache           *jwk.Cache
+	issuerToJwksURI *cache.Cache[string]
+}
+
+type OpenIDConfiguration struct {
+	// The only field we need from the config
+	JwksURI string `json:"jwks_uri"`
+}
+
+func NewOIDCResolver(ctx context.Context) *OIDCResolver {
+	return &OIDCResolver{
+		cache: jwk.NewCache(ctx),
+		// 24 hour cache timeout, jwks_uri should change very infrequently
+		issuerToJwksURI: cache.New[string](time.Hour * 24),
+	}
+}
+
+func (r *OIDCResolver) ResolveKeySet(ctx context.Context, oidcHeader string) (jwk.Set, error) {
+	issuer, err := r.extractIssuer(ctx, oidcHeader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract issuer from token: %w", err)
+	}
+
+	jwksURI, err := r.registerIssuer(ctx, issuer)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register issuer %q: %w", issuer, err)
+	}
+
+	return jwk.NewCachedSet(r.cache, jwksURI), nil
+}
+
+func (r *OIDCResolver) extractIssuer(ctx context.Context, oidcHeader string) (string, error) {
+	// Parse without validation to extract issuer so we can use it to obtain the key set to use for validation
+	token, err := jwt.ParseString(oidcHeader, jwt.WithContext(ctx), jwt.WithVerify(false))
+	if err != nil {
+		return "", fmt.Errorf("failed to parse jwt header: %w", err)
+	}
+	return token.Issuer(), nil
+}
+
+func (r *OIDCResolver) registerIssuer(ctx context.Context, issuer string) (string, error) {
+	jwksURI, err := r.resolveJwksURI(ctx, issuer)
+	if err != nil {
+		return "", fmt.Errorf("failed to resolve JWKS URI for %q: %w", issuer, err)
+	}
+
+	if !r.cache.IsRegistered(jwksURI) {
+		err = r.cache.Register(jwksURI)
+		if err != nil {
+			return "", fmt.Errorf("failed to register JWKS URI %q to cache: %w", jwksURI, err)
+		}
+	}
+
+	return jwksURI, nil
+}
+
+func (r *OIDCResolver) resolveJwksURI(ctx context.Context, issuer string) (string, error) {
+	uri, cached := r.issuerToJwksURI.Lookup(issuer)
+	if cached {
+		return uri, nil
+	}
+
+	configURL, err := url.JoinPath(issuer, ".well-known", "openid-configuration")
+	if err != nil {
+		return "", fmt.Errorf("error processing issuer URL %q: %w", issuer, err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, configURL, nil)
+	if err != nil {
+		return "", fmt.Errorf("error creating GET request for %q: %w", configURL, err)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("error fetching OpenID Configuration from %q: %w", configURL, err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
```

Comment:
Limit the max response size to something like 24MiB just to be safe.

Reply:
Done